server/routerlicious/packages/routerlicious-base/src/utils/sessionHelper.ts

```diff
@@ -134,20 +138,53 @@ async function updateExistingSession(
         isSessionActive: false,
     };
     try {
-        await documentsCollection.upsert(
+        const result = await documentsCollection.findAndUpdate(
             {
+                tenantId,
                 documentId,
+                "session.isSessionAlive": false,
             },
             {
+                createTime: document.createTime,
                 deli: updatedDeli ?? document.deli,
-                scribe: updatedScribe ?? document.scribe,
+                documentId: document.documentId,
                 session: updatedSession,
-            },
-            null);
+                scribe: updatedScribe ?? document.scribe,
+                tenantId: document.tenantId,
+                version: document.version,
+            });
         Lumberjack.info(
-            `The Session ${JSON.stringify(updatedSession)} was updated into the document collection`,
+            `The original document in updateExistingSession: ${result}`,
             lumberjackProperties,
         );
+        // There is no document with isSessionAlive as false. It means this session has been discovered by
+        // another call, and there is a race condition with different clients writing truth into the cosmosdb
+        // from different clusters. Thus, let it get the truth from the cosmosdb with isSessionAlive as true.
+        if (!result.existing) {
+            Lumberjack.info(`The document with isSessionAlive as false does not exist`, lumberjackProperties);
+            const doc: IDocument = await runWithRetry(
+                async () => documentsCollection.findOne({ tenantId, documentId, "session.isSessionAlive": true,}),
+                "getDocumentWithAlive",
+                3 /* maxRetries */,
+                1000 /* retryAfterMs */,
+                getLumberBaseProperties(documentId, tenantId),
+                undefined, /* shouldIgnoreError */
+                (error) => true, /* shouldRetry */
+            );
+            if (!doc && !doc.session) {
+                Lumberjack.error(
+                    `Error running getSession from document: ${JSON.stringify(doc)}`,
+                    lumberjackProperties,
+                );
+                throw new Error(`Error running getSession, please try again`);
```

Comment:
Is this an error that would be returned to the client? Should we use a `NetworkError` here instead and decide on a status code to return?

Reply:
Yes. I change it to NetworkError and set the status code to 500. Do you think it is reasonable? Thanks.

Reply:
Thanks @tianzhu007 ! I think that is reasonable. @pradeepvairamani , do you agree? Just in case Joel and I missed something.

Reply:
Looks good to me!